File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if(head->next==NULL && n==1) return NULL;
        else if(head->next==NULL && n==0) return head;

        ListNode* temp=head;
        int count=0;
        while(temp!=NULL){
            count++;
            temp=temp->next;
        }

        if(count==n){
            head=head->next;
            return head;
        }
        int k=count-n;
        temp=head;
        for(int i=1;i<k;i++){
            temp=temp->next;
        }

        temp->next=temp->next->next;
        return head;
    }
};

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode temp=head;
        int size=0;
        if(n==0) return head;
        while(temp!=null){
            temp=temp.next;
            size++;
        }
        if(size==n){
            head=head.next;
            return head;
        }
        int x=(size-n);
        temp=head;
        while(temp!=null && x>1){
            temp=temp.next;
            x--;
        }
        temp.next=temp.next.next;
        return head;
    }
}